src/Koffee/Utility.fs

```diff
@@ -36,6 +36,40 @@ module Format =
         else if size > scale 1 then scaledStr size 1
         else scaledStr size 0
 
+    type ISubstitutionProvider =
+        abstract member getSubstitution : key:string -> string option
+
+    type EnvSubstitutionProvider() =
+        static member instance = new EnvSubstitutionProvider()
+
+        interface ISubstitutionProvider with
+            member _.getSubstitution key =
+                Environment.GetEnvironmentVariable(key, EnvironmentVariableTarget.Process)
+                |> Option.ofObj
+
+    let private subRegex = new Regex(@"%([^%]+)%", RegexOptions.Compiled)
+
+    /// Substitutes variables in string with percent syntax.
+    /// Example input: "Hello %target%!"
+    /// Outputs: "Hello world!"
+    /// Where the "target" maps to "world"
+    let subVars (subs: ISubstitutionProvider) str =
+        let mayContainSubstitution (str: string) =
+            str.IndexOf '%' <> -1
+
+        let matchEval (m: Match) =
+            subs.getSubstitution m.Groups.[1].Value
+            |> Option.defaultValue m.Value
+
+        if mayContainSubstitution str then
+            subRegex.Replace(str, matchEval)
```

Comment:
Instead of only using the regex to replace, you can use it to check for the match too with `IsMatch`. Then you don't need the `mayContainSubstitution`

Reply:
I think after this refactor, this function will be quite short and could be moved to the controller. The tests can still test this specific function directly

Reply:
I used IndexOf mostly because it's much faster of a check. Probably premature optimization but switching to IsMatch would probably be slower than to just only call Regex.Replace all together and skipping the if-statement.

Can rename it to `fastCheckMayContainSub` to emphasize it, or replace the entire if-statement with `subRegex.Replace`. What do you think?